### ISSUE TITLE
sleep: system still hibernates instantly when using suspend-then-hibernate even when `on_ac_power()` returns true

### ISSUE THREAD

**Issue by axet** [systemd/systemd#25917]
### Component

systemd

### Is your feature request related to a problem? Please describe

Hello!

Right now systemd hibernates the suspended system on low battery (<5%) and external power connected.

It does not look like intended behavior since it would meant you expect user to disconnect power immediately after suspend (closing laptop lid) and to prevent data lose due to low battery you hibernate it before external power got disconnected. Seems like overkill. While external power is connected we should not hibernate the system. Battery can be charged or even report incorrect charge level due to age.

### Describe the solution you'd like

I suggest to check external power status before going to hibernate on low battery event. If external power is connected then do not hibernate and keep system in suspend state. Only when external power got disconnected check battery level and hibernate it.

EDIT:

### Workaround

_**/etc/modprobe.d/blacklist.conf**_
blacklist battery

**Comment by YHNdnzj** [comment]
Should be closed by https://github.com/systemd/systemd/pull/26503.

**Comment by axet** [comment]
@yuwata 

For some reason it got worse. Even removing 'battery' module no longer works. I have two system run with no battery installed. Systemd force hibernate all the time when 'syspend-then-hibernate' target used.

systemd                                      252.6-1

Looks like a bug in the code:

https://github.com/systemd/systemd-stable/blob/v252-stable/src/shared/sleep-config.c

Check on_ac_power call. It have been checked agains r > 0. But man page says:


0 (true)
    System is on line power. 
1 (false)
    System is not on line power. 
255 (false)
    Power status could not be determined.

**Comment by YHNdnzj** [comment]
Oh... https://salsa.debian.org/debian/powermgmt-base/-/blob/master/on_ac_power

This is some Debian stuff and has nothing to do with the function in `udev-util`.

**Comment by axet** [comment]
Ok. Just try "steps to reproduce" it suppose to not to hibernate instantly. Maybe it is another issue...

**Comment by YHNdnzj** [comment]
You can use `systemd-ac-power -v` to check what systemd's `on_ac_power` actually reports. Maybe that help to narrow down the problem.

**Comment by axet** [comment]
It reports 'yes' (two of my machines tested)

```
admin@m3n:~$ /lib/systemd/systemd-ac-power -v
yes
```

**Comment by YHNdnzj** [comment]
Please set the log level of `systemd-suspend-then-hibernate.service` to `debug` using `systemctl edit`:

```ini
[Service]
Environment=SYSTEMD_LOG_LEVEL=debug
```

Then attach the logs here.

**Comment by axet** [comment]
I have 3 machines:

1) mbp62 (macbookpro) with no battery.
2) m3n (asus m3n) old battery with random 0% 3% 70% charge with 'battery' module removed.
3) vivobook (asus) working fine. but for test purpose I remove 'battery' and it's dependencies module here.

Here are vivobook log:

[vivo.log](https://github.com/systemd/systemd/files/10978575/vivo.log)

**Comment by YHNdnzj** [comment]
I'm suspecting something in the battery trip point alarm part, but not really familiar with it 🤔

@axet Could you please set `HibernateDelaySec=` in `/etc/systemd/sleep.conf` (pick any value you like) and try again? Thanks!

**Comment by axet** [comment]
It is already:

HibernateDelaySec=16h

**Comment by axet** [comment]
PS: Machines with no battery installed or machines without 'battery' module hibernates instantly (maybe also machines with battery reporting 0%, this one hard to test)

**Comment by YHNdnzj** [comment]
Hopefully #26854 will fix this :)

**Comment by axet** [comment]
Thanks! Maybe I did not apply patch properly, but it keep hibernates instantly if here is no 'battery' module. Just checked one machine...

```
dpkg -i libsystemd-shared_252.6-1_amd64.deb  libsystemd0_252.6-1_amd64.deb  systemd_252.6-1_amd64.deb
```

**Comment by axet** [comment]
Reverting https://github.com/systemd/systemd/commit/4f58b656d92b09a953b7cffcfd1ee6d5136a57ed.patch indeed helps!

```
 patch -R -p1 < /media/axet/128GB/4f58b656d92b09a953b7cffcfd1ee6d5136a57ed.patch 
dpkg -i libsystemd-shared_252.6-1_amd64.deb  libsystemd0_252.6-1_amd64.deb  systemd_252.6-1_amd64.deb
```

**Comment by YHNdnzj** [comment]
> Maybe I did not apply patch properly, but it keep hibernates instantly if here is no 'battery' module. Just checked one machine...

Apparently the patch I submitted was spurious and tried to fix a problem that didn't exist...

**Comment by axet** [comment]
Since low battery level not common occasion I do not experiencing it often. Especially when my battery do not charge properly (charge level jumps from 0 to 80%). But now, fortunately, I got my battery level at 0 and found different behavior.

Using systemd 252.6-1 and low battery level, revering 4f58b65 does not help and system hibernates instantly. Although removing 'battery' mod from loaded kernel modules prevent system from hibernates instantly.

Summarize:

1) systemd 252.6-1 + revert-4f58b65, no 'battery' mod = suspend works fine
2) systemd 252.6-1 + revert-4f58b65, with 'battery' mod, battery 0%-5% = suspend does not works, hibernates instantly

**Comment by bluca** [comment]
@YHNdnzj do you intend to work on this for the v254 milestone?

**Comment by axet** [comment]
Lets do it slowly.

I have debian, and can only apply patched on top of 252.6-1. Last time I tried to build systemd from source my system went dead.

Can you point which patch are we talking about and I will apply it rebuild my systemd.

And I have to wait until my battery went low again.

**Comment by YHNdnzj** [comment]
> Can you point which patch are we talking about and I will apply it rebuild my systemd.

https://github.com/systemd/systemd/commit/e0b3a70fabb871bf55678e9e177445b1df2aee88

I'll remove this from the milestone then. Since `on_ac_power()` returns true for you, I'm pretty sure applying this will fix the problem.
